### DISCUSSION TITLE
Are Handlers intended to be returned?

### DISCUSSION THREAD

**Discussion by mschuchard** [XAMPPRocky/octocrab#411] [Q&A] [answered]
I have a question about the intended usage of the Octocrab handlers. I want to create a function for wrapping the `IssueHandler<'_>` constructor (PoC below):

```rust
pub fn new_issues<'octo>(pat: Option<String>, owner: &str, repo: &str) -> Result<octocrab::issues::IssueHandler<'octo>, ()> {
    // instantiate client
    let client = match pat {
        Some(pat) => octocrab::Octocrab::builder()
        .personal_token(pat)
        .build()
        .unwrap(),
        None => octocrab::Octocrab::default(),
    };
    // initalize and return issues
    return Ok(*client.issues(owner, repo))
}
```

Of course since `IssueHandler<'_>` does not impldement `Deref` I cannot seem to avoid the E0515 issue. This leads me to believe that the intended usage for Handler is that they should not be returned.

Is there any documentation that I can be directed towards or added to Octocrab about their intended usage to confirm or deny the intention here? If they are not intended to be returned, is there documentation for something similar in Octocrab that can be returned for functionality analogous to the above?

(documentation label implied for this issue)

**Comment by XAMPPRocky** [comment]
Thank you for your issue! I moved this to discussions, since it is more of a question.

Handlers are meant to be short-lived, you're supposed to consume them immediately. I'm not sure what you're trying to do with the `Deref`, but you'll also run into lifetime issues as the octocrab client you construct in the match statement needs to outlive the handler.

Can you provide more detail on what you're trying to achieve?

**Comment by mschuchard** [comment]
Indeed a question, and this is the first repo I have encountered with a "discussions" so I was unfamiliar with it.

I am attempting to construct a general `octocrab::issues::IssueHandler<'octo>` and then utilize it for specific actions related to the Issues API by passing it into other functions as a pointer. This is a common design pattern for SDK bindings usage in other languages, and so I was attempting to copy that pattern here (perhaps sub-optimally).

If I cannot return it from a wrapped constructor, then I assume that I would need to either reconstruct the same object repeatedly (inefficient), or wrap both the constructor and the other Issue API functions within a general function (awkward usage).

Is there some better design pattern here?

**Reply by XAMPPRocky** [reply, marked as answer]
Well I'm not sure about passing the function as a pointer, but I would probably use closures instead. If you know what the return type will be, you can replace `I` with that type.

```rust
pub fn new_issues<I, F>(
    pat: Option<String>, 
    owner: &str, 
    repo: &str,
    customise: impl FnOnce(octocrab::issues::IssueHandler) -> F
) -> Result<I, ()>
    where F: Future<Output = Result<I, ()>>
{
    // instantiate client
    let client = match pat {
        Some(pat) => octocrab::Octocrab::builder()
        .personal_token(pat)
        .build()
        .unwrap(),
        None => octocrab::Octocrab::default(),
    };
    // initalize and return issues
    return (customise)(client.issues(owner, repo)).await
}
```

```rust
let issue: Issue = new_issues(None, "XAMPPRocky", "octocrab", |handler| {
      handler.get(1)
}).await.unwrap();
```

**Reply by mschuchard** [reply]
Ok that solution is definitely too clever for someone like me with my too broad/unspecialized of a skill set. I understand the generic can totally be replaced, but using a lambda/closure like that as a parameter to maintain the lifetime is something I probably never would have figured out. I do recognize that this is totally a Rust design pattern so I clearly need to be in that "headspace" more.

Please allow me to progress a bit more as I am still in the PoC phase so I am unsure yet if this completely solves my use case or not. Thank you.

**Reply by XAMPPRocky** [reply]
> using a lambda/closure like that as a parameter to maintain the lifetime is something I probably never would have figured out

No worries, everyone has to learn at some point. In general if you need to control the lifecycle of an operation, for example, if you need to run some code after the caller operations finishes. A closure is the only way in Rust to do that for synchronous operations (async is a bit different with futures, but the principle is the same.)

**Reply by mschuchard** [reply]
Ok I put some more effort into this and have some takeaways:
- I refactored from robust functionality into limited and simple usage direct from the documentation examples and am still slamming into Rust issues preventing compilation, and so I will need to ensure the usage is as limited and simple as possible if I am developing my plugin with Rust instead of another language.
- I respect this SDK a lot more now and understand better why so few are developed in Rust. Conversely I am unsure why so many are developed in Go and Python, but it appears I will need to continue developing heavily in those two languages for some time to come with all the SDK work I do.
- This experience reminds me of Go several years ago, so I would definitely return to a Rust SDK in the future since Go eventually worked itself out. On a personal note I see Rust more as a successor to Go anyway than C++ (although the power and complexity are similar to me) and do not comprehend all the arguments in the Rust vs. C++ space too.

Thanks again for the help. I will definitely continue attempting to develop the plugin with this SDK but in a much more limited capacity.

Also I totally understand where  GH discussions are better suited than SO Q&A so thanks for enabling this in your repo.

**Reply by XAMPPRocky** [reply]
I don't have the context of what you're trying to make, but just from your previous example, I think you're making it harder for yourself by creating the client inside the function, you'd have a lot easier time exposing the client directly, or by wrapping it in a `Context` object or similar and have your users pass that around to your SDK functions.

**Reply by mschuchard** [reply]
So actually I somehow missed the [Static API](https://docs.rs/octocrab/latest/octocrab/#static-api) in the docs, and that has analogous functionality to what I was seeking. You were already providing what I needed it seems.

Regardless, I still observe various compilation errors for all verbatim examples from the Octocrab code and docs for version 1.66.1 and edition 2021, so unfortunately I will likely have to implement the plugin with the Github Go SDK instead (Go finally becoming good with 1.18 last year). I really wanted to use Rust for this use case though since I am developing a smaller and simpler plugin, but it appears I will need to revisit Rust in 2024.

Thanks again for the assistance.

**Reply by XAMPPRocky** [reply]
>  I still observe various compilation errors for all verbatim examples from the Octocrab code

If you're getting compilation issues from the examples, then that's a bug, but I think you're doing something wrong as while the examples don't test the run time, we do test if they compile, and they all do because that's part of CI. Could you provide an example of the kind of error you're seeing?

**Reply by mschuchard** [reply]
Yeah I just figured out how to bypass the Rust issues here and I feel like I knew it for a while and was just avoiding it: this crate has a hard requirement on `tokio`. Adding the attribute macro `#[tokio::main]` (after searching in various places for the magical Cargo config to enable it) to functions fixes the errors and everything works with it. Without it, any usage of the crate either from examples, code examples, or doc examples (mysteriously whether `async` or not) fails to compile (usually with typing issues). While I suspected as much this was the root cause I am unsure how obvious that should have been to me. I do not feel like even C++ has this kind of ramp up.

So basically with that one line of code all of my code now suddenly functions correctly. I assume it should be obvious since it is absent from the documentation, so this is probably due to my lack of experience with concurrency as opposed to multi-process/multi-threading (where I do have familiarity).

**Reply by XAMPPRocky** [reply]
Yeah, unlike Go, Rust does not currently include an async runtime, so in order to be able to use async code, you need to use a runtime like Tokio.

I am gradually moving this library away from depending on Tokio, but you'd still need to choose a runtime even if/when that dependency is removed.
